File: frontend/src/types/goal.ts
export interface GoalBase {
    name: string;
    target_amount: number;
    target_date: string; // YYYY-MM-DD
    current_savings: number;
    monthly_contribution: number;
    risk_profile?: string;
}

export interface SimulationRequest extends GoalBase { }

export interface YearlyProjection {
    year: number;
    amount: number;
    contribution: number;
    growth: number;
}

export interface ScenarioProjections {
    median: YearlyProjection[];
    worst: YearlyProjection[];
    best: YearlyProjection[];
}

export interface ETFDetail {
    ticker: string;
    name: string;
    percent: number;
    desc: string;
}

export interface SimulationResult {
    success_probability: number;
    gap_amount: number;
    projected_amount_at_deadline: number;
    recommended_allocation: { [key: string]: number };
    etf_suggestions: { [key: string]: ETFDetail };
    yearly_projections: YearlyProjection[];
    scenarios: ScenarioProjections;
    on_track: boolean;
    suggestions: { [key: string]: number };
}

File: backend/app/schemas/goal.py
from pydantic import BaseModel
from typing import List, Dict, Optional, Any
from datetime import date

class GoalBase(BaseModel):
    name: str
    target_amount: float
    target_date: date
    current_savings: float = 0.0
    monthly_contribution: float = 0.0
    risk_profile: Optional[str] = "Moderate"

class GoalCreate(GoalBase):
    pass

class GoalResponse(GoalBase):
    id: int
    created_at: date

class YearlyProjection(BaseModel):
    year: int
    amount: float
    contribution: float
    growth: float

class ScenarioProjections(BaseModel):
    median: List[YearlyProjection]
    worst: List[YearlyProjection]
    best: List[YearlyProjection]

class SimulationResult(BaseModel):
    success_probability: float  # 0 to 1
    gap_amount: float # Amount needed to reach goal if behind
    projected_amount_at_deadline: float
    recommended_allocation: Dict[str, float] # e.g. {"Equity": 0.7, "Bonds": 0.3}
    etf_suggestions: Dict[str, Any] # Detailed ETF info
    yearly_projections: List[YearlyProjection] # Deprecated but kept for compatibility
    scenarios: ScenarioProjections
    on_track: bool
    suggestions: Dict[str, float]

class SimulationRequest(GoalBase):
    pass
